tapchannel/aux_invoice_manager_test.go

```diff
@@ -0,0 +1,571 @@
+			},
+		},
+		{
+			name: "asset invoice, custom records",
+			requests: []lndclient.InvoiceHtlcModifyRequest{
+				{
+					Invoice: &lnrpc.Invoice{
+						RouteHints: testRouteHints(),
+						ValueMsat:  3_000_000,
+					},
+					WireCustomRecords: newWireCustomRecords(
+						t, []*rfqmsg.AssetBalance{
+							rfqmsg.NewAssetBalance(
+								assetID(1), 3,
+							),
+						}, fn.Some(dummyRfqID(31)),
+					),
+				},
+			},
+			responses: []lndclient.InvoiceHtlcModifyResponse{
+				{
+					AmtPaid: 3_000_000,
+				},
+			},
+			buyQuotes: rfq.BuyAcceptMap{
+				dummyRfqID(31).Scid(): {
+					Peer:     testNodeID,
+					AskPrice: 1_000_000,
+				},
+			},
+		},
+		{
+			name: "asset invoice, not enough amt",
+			requests: []lndclient.InvoiceHtlcModifyRequest{
+				{
+					Invoice: &lnrpc.Invoice{
+						RouteHints: testRouteHints(),
+						ValueMsat:  3_000_000,
+					},
+					WireCustomRecords: newWireCustomRecords(
+						t, []*rfqmsg.AssetBalance{
+							rfqmsg.NewAssetBalance(
+								assetID(1), 3,
+							),
+						}, fn.Some(dummyRfqID(31)),
+					),
+					ExitHtlcAmt: 1234,
+				},
+			},
+			responses: []lndclient.InvoiceHtlcModifyResponse{
+				{
+					AmtPaid: 1_500_000,
+				},
+			},
+			buyQuotes: rfq.BuyAcceptMap{
+				dummyRfqID(31).Scid(): {
+					Peer:     testNodeID,
+					AskPrice: 500_000,
+				},
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+
+		t.Logf("Running AuxInvoiceManager test case: %v", testCase.name)
+
+		// Instantiate mock rfq manager.
+		mockRfq := &mockRfqManager{
+			peerBuyQuotes:   testCase.buyQuotes,
+			localSellQuotes: testCase.sellQuotes,
+		}
+
+		done := make(chan bool)
+
+		// Instantiate mock htlc modifier.
+		mockModifier := &mockHtlcModifier{
+			requestQue:     testCase.requests,
+			expectedResQue: testCase.responses,
+			done:           done,
+			t:              t,
+		}
+
+		// Create the manager.
+		manager := NewAuxInvoiceManager(
+			&InvoiceManagerConfig{
+				ChainParams:         testChainParams,
+				InvoiceHtlcModifier: mockModifier,
+				RfqManager:          mockRfq,
+			},
+		)
+
+		err := manager.Start()
+		require.NoError(t, err)
+
+		// If the manager is not done processing the htlc modification
+		// requests within the specified timeout, assume this is a
+		// failure.
+		select {
+		case <-done:
+		case <-time.After(testTimeout):
+			t.Fail()
+		}
+	}
+}
+
+// genRandomRfqID generates a random rfqmsg.ID value.
+func genRandomRfqID(t *rapid.T) rfqmsg.ID {
+	var rfqID rfqmsg.ID
+	for i := 0; i < len(rfqID); i++ {
+		rfqID[i] = rapid.Byte().Draw(t, "scid_byte")
+		// rapid.SliceOfN(rapid.Byte(), len(rfqID), len(rfqID))
+	}
+	return rfqID
+}
+
+// genInvoice generates an invoice that may have a random amount, and may have
+// routing hints.
+func genInvoice(t *rapid.T) *lnrpc.Invoice {
+	res := &lnrpc.Invoice{}
+
+	// Do a random draw with a 1/8 chance of introducing an empty invoice.
+	nullInv := rapid.Uint16().Draw(t, "invoice_exists")
```

Comment:
I think the places where you to pick a random, probably invalid, value with some frequency, we can just use `rapid.Bool()` instead.

Also, if you separate the Generator for `lnrpc.Invoice` into another function, ideally that matches the `Generator` function signature:

https://pkg.go.dev/pgregory.net/rapid#Custom

Then a generator for `*lnrpc.Invoice` using https://pkg.go.dev/pgregory.net/rapid#Ptr is straightforward:

```go
func invoiceGen(t *rapid.T) lnrpc.Invoice {
/// implementation
}

func requestGen(t *rapid.T) (.....) {
/// implementation 

request.Invoice = rapid.Ptr(invoiceGen, true).Draw(t, "request invoice")

/// implementation
}
```

Reply:
Given the many fields that **may** be invalid, we would be gravitated towards spending the majority of the test runs in early failures.

For example, if the above `"invoice_exists"` draw was given a 50% chance, we'd effectively dedicate 50% of our total runs into the same invoice check in the code.

Reply:
I will expose all those "magic numbers" that somehow express probability of smth happening as top level const values, so it's easier to inspect and change

Reply:
> For example, if the above `"invoice_exists"` draw was given a 50% chance, we'd effectively dedicate 50% of our total runs into the same invoice check in the code.

I don't think that's true; the framework should determine that this variable causes early failures and then leave this as `non-nil` for future iterations, after some iterations where this was `nil`.

Are you observing an issue with this version in practice?

Reply:
I agree, we should let the framework decide what it "spends time on". That seems to be the whole purpose that it knows how to most efficiently discover code paths.